### PR TITLE
docker: Add storage panel

### LINKED ISSUE #1662: Need a good and easy way to tell the size and space avaiable to /var/lib/docker
Spinning off issue #1541

We need a good way to tell the total size available to /var/lib/docker to support the use cases of the personas Peter and David here:
https://github.com/cockpit-project/cockpit/wiki/Atomic:-Docker-Storage

### PR DESCRIPTION
**Pull request by petervo**
Fixes #1662 

The only graphdriver that reports any usage statistics is devicemapper. In a default setup that will use a sparse file as the loopback device for it's pool. Docker reports the size of that file as the available space even if the underlying device really doesn't have that much space available. I'm going to see if they are interested in reporting the actual available space but in the mean time this just has a disclaimer.

Currently this is only updating on load, triggering it on events seemed a little overkill but if we think that it should be fresher maybe we should use a timer.

### CODE REVIEW

File: pkg/shell/cockpit-docker.js [hunk 1]
```diff
@@ -605,6 +606,47 @@ PageContainers.prototype = {
...
+        this.client.info().done(function(data) {
+            var resp = data && JSON.parse(data);
+            if (resp['Driver'] !== "devicemapper") {
+                // TODO: None of the other graphdrivers currently
+                // report size information.
+                $('#containers-storage .bar').html();
+                $('#containers-storage .data').html("Unknown");
+            }
+
+            var warning = "WARNING: Docker may be reporting the size it has alocated to it's storage pool using sparse files, not the actual space available to the underlying storage device.";
```
Comment (stefwalter) [inline pkg/shell/cockpit-docker.js]: Lets make this translatable with ```_(...)```

File: pkg/shell/cockpit-docker.js [hunk 2]
```diff
@@ -605,6 +606,47 @@ PageContainers.prototype = {
...
+            $('#containers-storage').attr("title", warning).tooltip();
+
+            var used;
+            var total;
+            $.each(resp['DriverStatus'], function (index, value) {
+                if (value && value[0] == "Data Space Total")
+                    total = value[1];
+                else if (value && value[0] == "Data Space Used")
+                    used = value[1];
```
Comment (stefwalter) [inline pkg/shell/cockpit-docker.js]: Out of interest, does Docker translate these strings?

File: pkg/shell/cockpit-docker.js [hunk 3]
```diff
@@ -585,6 +585,7 @@ PageContainers.prototype = {
         }
 
         setup_for_failure(self, self.client, self.address);
+        self.render_storage();
```
Comment (stefwalter) [inline pkg/shell/cockpit-docker.js]: How often do we render the storage space? Is it worth updating it?

**Comment by stefwalter** [comment]
Nice. Looks good and seems to work well. Obviously once we have fixed docker to report more relevant info, this will be even more useful.

**Comment by stefwalter** [comment]
Seems like the hubbot failure here is a false positive. @mvollmer is that right?

**Comment by mvollmer** [comment]
> Seems like the hubbot failure here is a false positive. @mvollmer is that right?

Yep.  It is at least a known and unrelated failure.  I start to suspect that more might be wrong than just the test.

**Comment by petervo** [comment]
Should we be exposing metadata space as well?

**Comment by petervo** [comment]
Like i said above It's only updating on load, i had hooked it up to events but was was generating a lot of calls and seemed overkill for something that updates pretty slowly. Should I put it on a timer?

**Comment by petervo** [comment]
Docker doesn't seem to translate those string, Marked the warning as translatable.

**Comment by petervo** [comment]
Added some code to fix the graph when the units for used and total are different.

**Comment by stefwalter** [comment]
> Should we be exposing metadata space as well?

Can we leave this out for now? Once we're reporting the file system space on the outside (with your docker patches), then I think that'll get us closer to representing the fact that space is running low for *both* the data and metadata loopback files.

**Comment by stefwalter** [comment]
> Like i said above It's only updating on load, i had hooked it up to events but was was generating a lot of calls and seemed overkill for something that updates pretty slowly. Should I put it on a timer?

How about a rate limited, triggered by events. That is when the first event comes in we reload the info, but then for the next event, if it comes in within 10 seconds of the first, wait at least 10 seconds etc.?

**Comment by petervo** [comment]
I'm good with leaving out meta for now. 

Added triggering graph updates with events.

File: pkg/base/cockpit.js [hunk 4]
```diff
@@ -2274,6 +2274,25 @@ function full_scope(cockpit, $, po) {
         /* 1024: [ null, "KiB", "MiB", "GiB", "TiB", "PiB", "EiB", "ZiB" ] */
     };
 
+    cockpit.bytes_from_format = function bytes_from_format(formatted, factor, separate) {
```
Comment (stefwalter) [inline pkg/base/cockpit.js]: Isn't this docker specific? Can we leave it in docker/docker.js for now? Or do you have plans to use this somewhere else? In general "screen scraping" is unfortunate and usually a work around for a junky API. Ideally we won't be using this code in many places? Or am I wrong?

Comment (petervo) [inline pkg/base/cockpit.js]: I hope you're right, i'll move.

**Comment by petervo** [comment]
Moved bytes_from_format.

**Pull request closed without merging**